features/distributed/kv-store/postgres-shared/src/main/java/org/opennms/features/distributed/kvstore/pgshared/AbstractPostgresKeyValueStore.java

```diff
@@ -0,0 +1,250 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.distributed.kvstore.pgshared;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.concurrent.TimeUnit;
+
+import javax.sql.DataSource;
+
+import org.opennms.features.distributed.kvstore.api.AbstractAsyncKeyValueStore;
+
+/**
+ * A {@link org.opennms.features.distributed.kvstore.api.KeyValueStore} backed by Postgres.
+ * <p>
+ * Postgres key value stores should implement this class with concrete types.
+ *
+ * @param <T> the type this store persists
+ * @param <S> the SQL type this store persists if the value type must be wrapped with an SQL type
+ */
+public abstract class AbstractPostgresKeyValueStore<T, S> extends AbstractAsyncKeyValueStore<T> {
+    private static final String VALUE_COLUMN = "value";
+    private static final String KEY_COLUMN = "key";
+    private static final String CONTEXT_COLUMN = "context";
+    private static final String LAST_UPDATED_COLUMN = "last_updated";
+    private static final String EXPIRES_AT_COLUMN = "expires_at";
+
+    private final PreparedStatement selectStatement;
+    private final PreparedStatement upsertStatement;
+    private final PreparedStatement lastUpdatedStatement;
+
+    public AbstractPostgresKeyValueStore(DataSource dataSource) {
+        Objects.requireNonNull(dataSource);
+
+        try {
+            Connection connection = dataSource.getConnection();
```

Comment:
This connection should eventually be closed. There's also a chance of it going "stale", so we'll need to do some better connection management here.